server/PathServices/SignalRRealtimeDataRepository.cs

```diff
@@ -0,0 +1,236 @@
+
+        private IEnumerable<RealtimeData> GetRealtimeData(Station station, RouteDirection direction)
+        {
+            return this.realtimeData.GetValueOrDefault((station, direction), new List<RealtimeData>());
+        }
+
+        private void PathSqlDbUpdated(object sender, EventArgs args)
+        {
+            Log.Logger.Here().Information("Recreating SignalR hubs following a PATH DB update...");
+            Task.Run(this.CreateHubConnections).Wait();
+        }
+
+        private async Task CreateHubConnections()
+        {
+            await this.CloseExistingHubConnections();
+
+            var tokenBrokerUrl = Decryption.Decrypt(await this.pathDataRepository.GetTokenBrokerUrl());
+            var tokenValue = Decryption.Decrypt(await this.pathDataRepository.GetTokenValue());
+
+            await Task.WhenAll(StationMappings.StationToSignalRTokenName.SelectMany(station =>
+                RouteDirectionMappings.RouteDirectionToDirectionKey.Select(direction => this.CreateHubConnection(tokenBrokerUrl, tokenValue, station.Key, direction.Key))));
+        }
+
+        private async Task CreateHubConnection(string tokenBrokerUrl, string tokenValue, Station station, RouteDirection direction, int sequentialFailures = 0)
+        {
+            SignalRToken token;
+
+            try
+            {
+                Log.Logger.Here().Information("Creating new SignalR hub connection for S:{station} D:{direction}", station, direction);
+                token = await this.pathApiClient.GetToken(tokenBrokerUrl, tokenValue, station, direction);
+
+                var connection = new HubConnectionBuilder()
+                    .WithUrl(token.Url, c => c.AccessTokenProvider = () => Task.FromResult(token.AccessToken))
+                    .Build();
+
+                connection.On<string, string>("SendMessage", (_, json) =>
+                    this.ProcessNewMessage(station, direction, json)
+                        .ConfigureAwait(false)
+                        .GetAwaiter()
+                        .GetResult());
+
+                async Task RetryConnection()
+                {
+                    await Task.Delay(new Random().Next(1, 7) * (1000 * Math.Min(sequentialFailures + 1, 5)));
+                    await this.CreateHubConnection(tokenBrokerUrl, tokenValue, station, direction, sequentialFailures + 1);
+                };
+
+                connection.Closed += async (e) =>
+                {
+                    if (!this.hubConnections.ContainsKey((station, direction)))
+                    {
+                        return;
+                    }
+
+                    if (e != null)
+                    {
+                        Log.Logger.Here().Warning(e, "SignalR connection was closed as a result of an exception");
+                    }
+
+                    Log.Logger.Here().Information("Recovering SignalR connection to {station}-{direction}...", station, direction);
+                    await RetryConnection();
```

Comment:
Should we short-circuit out of here whenever we need to RetryConnection? (both here and in the catch block below)